### PR TITLE
Use isolated self-hosted runners non-PR CI jobs

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

An attack may send an PR to try to install an improper code on self-hosted runners after sending a normal PR.

GitHub Actions only allows non-forked workflows to access secrets. If we isolate self-hosted runners for non-forked workflows, the vulnerable code in the attacker's PR can't acquire sensitive information.

Modifications:

- Split `self-hosted` runners into `self-hosted-for-pr` and `self-hosted-for-non-pr`.
  - `self-hosted-for-non-pr` is run with GitHub Actions secrets.
  - `self-hosted-for-pr` is used for CI builds of pull requests.
- `check-workflow-push-permission.sh` is not directly used in GitHub workflow but is triggered by the GitHub Actions runner before starting a job

Result:

Respond to possible vulnerabilities by using self-hosted.

### CODE REVIEW

**Review by minwoox — APPROVED** [review]
Thanks!
I don't fully understand this but let me ask you tomorrow. 😉

**Comment by ikhoon** [comment]
When a non-maintainer changes workflow files, the workflows will be rejected from execution: https://github.com/line/armeria/actions/runs/7610334781/job/20723579232?pr=5402#step:2:5
When a maintainer changes workflow files, the workflows are allowed to run: https://github.com/line/armeria/actions/runs/7610692738/job/20724558885?pr=5402#step:2:5

**Comment by ikhoon** [comment]
This PR is now ready to review.

File: .github/check-workflow-write-permission.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+if [[ $GITHUB_REF != refs/pull* ]]; then
+  echo "Not a pull request. Skipping"
+  exit 0
+fi
+
+PR_NUMBER=$(echo "$GITHUB_REF" | awk -F / '{print $3}')
+# To obtain sufficient quota for gh cli, it is recommended to set PAT in the `GH_TOKEN` environment variable.
+# Check if there are any changes in .github/actions or .github/workflows
+WORKFLOW_CHANGES=$(gh -R line/armeria pr diff "$PR_NUMBER" --name-only | grep -c '^.github/workflows\|^.github/actions')
```
Comment (jrhee17) [inline .github/check-workflow-write-permission.sh:17]: Question) Is it possible for external contributor change this script itself (which isn't in `workflows`, `actions`)

File: .github/workflows/publish-release.yml [hunk 2]
```diff
@@ -12,7 +12,7 @@ jobs:
   publish:
     name: Publish final artifacts
     if: github.repository == 'line/armeria'
-    runs-on: self-hosted
+    runs-on: self-hosted-for-non-pr
```
Comment (jrhee17) [inline .github/workflows/publish-release.yml:15]: optional; `self-hosted-safe`, `self-hosted-unsafe` might be more straightforward

**Review by jrhee17 — APPROVED** [review]
Left some minor questions but looks good! Thanks @ikhoon 🙇 👍 🙇

File: .github/check-workflow-write-permission.sh (hunk 1 above)
Comment (ikhoon) [reply .github/check-workflow-write-permission.sh:17]: The file is not directly accessed by GitHub Actions. It was added for version control. If there are any changes to this file, we must manually update the file in self-hosted runners.

**Comment by ikhoon** [comment]
Thanks for the quick review. ❤️

**Pull request merged**